### PR TITLE
Add mod dependencies for minor optimisations of prime fields

### PR DESCRIPTION
**Pull request by Multramate**
### Pull Request Checklist

- [x] Update [ChangeLog.md](https://github.com/adjoint-io/galois-field/blob/master/ChangeLog.md)
- [x] Tests are passing.
- [x] Benchmarks are not regressing.

### CODE REVIEW

File: src/Data/Field/Galois/Prime.hs [hunk 1]
```diff
@@ -61,9 +62,7 @@ instance KnownNat p => GaloisField (Prime p) where
 instance KnownNat p => Group (Prime p) where
   invert = recip
   {-# INLINE invert #-}
-  pow y@(P x) n
-    | n >= 0    = P $ powModNatural x (fromIntegral n) $ natVal (witness :: Prime p)
-    | otherwise = pow (recip y) $ P.negate n
+  pow x  = P . unMod . (^%) (unsafeCoerce x :: Mod p)
```
Comment (Multramate) [inline src/Data/Field/Galois/Prime.hs]: In order to use the hidden constructor `Mod` indirectly, I had to add six instances of `unsafeCoerce`, which currently works but may cause some problems with more aggressive optimisations.

**Comment by sdiehl** [comment]
I think this is fine. What is the delta in performance with and without unsafeCoerce?

**Comment by Multramate** [comment]
> I think this is fine. What is the delta in performance with and without unsafeCoerce?

Without `unsafeCoerce`, we would have to use `fromInteger`, which does unnecessary slow modular reductions. Using the current benchmarks on my machine,
- addition takes ~45ns before this PR, ~35ns with `unsafeCoerce`, ~105ns with `fromInteger`, and
- multiplication takes ~145ns before this PR, ~135ns with `unsafeCoerce`, ~205ns with `fromInteger`.
I did not measure the other operations but their deltas are relatively negligible (e.g. division already takes ~1us). Essentially, there is a minor boost in performance in all fronts, but with higher robustness due to abstracting low-level optimisations completely to `mod`.

**Comment by sdiehl** [comment]
Ok sounds good to me.

**Pull request merged**